test/suites/integration/nodeclass_test.go

```diff
@@ -0,0 +1,182 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration_test
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	coretest "sigs.k8s.io/karpenter/pkg/test"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Nodeclass Validation", func() {
+	var createPolicyOutput *iam.CreatePolicyOutput
+	var roleName string
+	var err error
+
+	AfterEach(func() {
+		_, err := env.IAMAPI.DetachRolePolicy(env.Context, &iam.DetachRolePolicyInput{
+			RoleName:  aws.String(roleName),
+			PolicyArn: createPolicyOutput.Policy.Arn,
+		})
+		Expect(err).To(BeNil())
+
+		_, err = env.IAMAPI.DeletePolicy(env.Context, &iam.DeletePolicyInput{
+			PolicyArn: createPolicyOutput.Policy.Arn,
+		})
+		Expect(err).To(BeNil())
+	})
+	It("should fail reconciliation when a required permission is explicitly denied", func() {
+		createPolicyInput := &iam.CreatePolicyInput{
+			PolicyName: aws.String("DenyPolicy"),
+			PolicyDocument: aws.String(`{
+				"Version": "2012-10-17",
+				"Statement": [
+					{
+						"Effect": "Deny",
+						"Action": [
+							"ec2:CreateFleet",
+							"ec2:CreateLaunchTemplate",
+							"ec2:RunInstances"
+						],
+						"Resource": "*"
+					}
+				]
+			}`),
+		}
+
+		roleName = fmt.Sprintf("%s-karpenter", env.ClusterName)
+
+		createPolicyOutput, err = env.IAMAPI.CreatePolicy(env.Context, createPolicyInput)
+		Expect(err).To(BeNil())
+
+		_, err = env.IAMAPI.AttachRolePolicy(env.Context, &iam.AttachRolePolicyInput{
+			RoleName:  aws.String(roleName),
+			PolicyArn: createPolicyOutput.Policy.Arn,
+		})
+		Expect(err).To(BeNil())
+
+		pod := coretest.Pod()
+		env.ExpectCreated(nodePool, nodeClass, pod)
+
+		createdPod := &corev1.Pod{}
+		Eventually(func(g Gomega) {
+			err := env.Client.Get(env.Context, client.ObjectKey{
+				Namespace: pod.Namespace,
+				Name:      pod.Name,
+			}, createdPod)
+			g.Expect(err).To(BeNil())
+		}, "30s", "5s").Should(Succeed())
+
+		Consistently(func(g Gomega) {
+			err := env.Client.Get(env.Context, client.ObjectKeyFromObject(pod), pod)
+			g.Expect(err).To(BeNil())
+			g.Expect(pod.Spec.NodeName).To(Equal(""))
+		}, "30s", "5s").Should(Succeed())
+
+		Eventually(func(g Gomega) {
```

Comment:
Can we check the nodeclass status condition instead of events? The would be better to validate the changes we want have been made